### ISSUE TITLE
🥹black image after called rasterization in simple_viewer.py

### ISSUE THREAD

**Issue by Atticuszz** [nerfstudio-project/gsplat#243]
Hello !
thanks for your gorgeous project！
here I've encountered some problems  as  
 using 
```Python 
def rasterization(
    means: Tensor,  # [N, 3]
    quats: Tensor,  # [N, 4]
    scales: Tensor,  # [N, 3]
    opacities: Tensor,  # [N]
    colors: Tensor,  # [(C,) N, D] or [(C,) N, K, 3]
    viewmats: Tensor,  # [C, 4, 4]
    Ks: Tensor,  # [C, 3, 3]
    width: int,
    height: int,
    near_plane: float = 0.01,
    far_plane: float = 1e10,
    radius_clip: float = 0.0,
    eps2d: float = 0.3,
    sh_degree: Optional[int] = None,
    packed: bool = True,
    tile_size: int = 16,
    backgrounds: Optional[Tensor] = None,
    render_mode: Literal["RGB", "D", "ED", "RGB+D", "RGB+ED"] = "RGB",
    sparse_grad: bool = False,
    absgrad: bool = False,
    rasterize_mode: Literal["classic", "antialiased"] = "classic",
    channel_chunk: int = 32,
) -> Tuple[Tensor, Tensor, Dict]:
```

in rgb mode got almost black image 
which params should I check?

thank you!

**Comment by MatteoFusconi** [comment]
I've encountered a similar problem: I am manually editing a gaussian splat (produced with splatfacto new version), in particular adding some gaussians. When I run the rasterization, I notice that I will see the gaussians only if the camera point of view is close enough, whereas if it is far, they will not be rasterized, and I will see only the old ones.

However, I think it's a matter of some new feature of the rasterizer, since inria rasterizer works, and online viewers like https://antimatter15.com/splat/ works.

Can somebody give us better insights?
Thank you :)

**Comment by liruilong940607** [comment]
Hi,

For @MatteoFusconi 's question, I'm wondering if it is related to the `radius_clip` argument at here, which is to skip rendering the Gaussians that are very small in the image plane (< 3 pixels).
https://github.com/nerfstudio-project/gsplat/blob/0ac847d6efde31dba8c289c67241321a3cf9821f/examples/simple_viewer.py#L171-L172

That is set there by default to speedup rendering. Maybe you could try setting it to zero? Other than than I don't see any logical difference between `simple_renderer` and inria's rendering. And since your model is trained with splatfacto, which is using gsplat under the hood, the data should be very much compatible as well.

For @Atticuszz, what is exactly the way that you use `simple_viewer.py`? How is the model trained? Is the images ALMOST black or COMPLETELY black? Would you be able to see something if you move around the viewpoint in the viewer (e.g., zoom in or out?) A reasonable guess that I have is that if your data is not normalized to the center of the world coordinate system, you might need to "find" your scene in the viewer. The viewer by default looks at (0, 0, 0) when it opens up, it does not have any knowledge on where your scene might be.

**Comment by MatteoFusconi** [comment]
It should not be the case, as I used the SplatfactoModel.get_outputs method to produce the render, so it should have that line commented, and not use the radius_clip.

To provide additional context, I can say that the SuperSplat viewer behaves exactly like gsplat, whereas antimatter behaves like inria.

Thank you for your help :)

**Comment by liruilong940607** [comment]
> To provide additional context, I can say that the SuperSplat viewer behaves exactly like gsplat, whereas antimatter behaves like inria.

@MatteoFusconi Interesting to hear that. I would love to know more about it. Any data/code/info you could provide for us to look into it?

**Comment by Atticuszz** [comment]
> Hi,
> 
> For @MatteoFusconi 's question, I'm wondering if it is related to the `radius_clip` argument at here, which is to skip rendering the Gaussians that are very small in the image plane (< 3 pixels).
> https://github.com/nerfstudio-project/gsplat/blob/0ac847d6efde31dba8c289c67241321a3cf9821f/examples/simple_viewer.py#L171-L172
> 
> That is set there by default to speedup rendering. Maybe you could try setting it to zero? Other than than I don't see any logical difference between `simple_renderer` and inria's rendering. And since your model is trained with splatfacto, which is using gsplat under the hood, the data should be very much compatible as well.
> 
> For @Atticuszz, what is exactly the way that you use `simple_viewer.py`? How is the model trained? Is the images ALMOST black or COMPLETELY black? Would you be able to see something if you move around the viewpoint in the viewer (e.g., zoom in or out?) A reasonable guess that I have is that if your data is not normalized to the center of the world coordinate system, you might need to "find" your scene in the viewer. The viewer by default looks at (0, 0, 0) when it opens up, it does not have any knowledge on where your scene might be.

awesome！😄😄😄，i use the rgb-d data-replica，normalized  follow theoriginal code， test it with two image has 810000×2=1620000 gs ，so `radius_clip` skip too much，set it as 0.0 and scene grid as 1 it works for me 
thanks for your patient advics ☺️

**Comment by MatteoFusconi** [comment]
Goodmorning, I will provide some additional context.
I am rendering with a slightly modified version of the code for rendering datasets (ns-render dataset --split train)

However, the part where I obtain the output is still:
```
outputs = pipeline.model.get_outputs_for_camera(
                           cameras[camera_idx : camera_idx + 1], obb_box=obb_box
                     )
```

which, under the hood, calls the rasterization in this way (in file nerfstudio/models/splatfacto, in function get_outputs):
```
render, alpha, info = rasterization(
            means=means_crop,
            quats=quats_crop / quats_crop.norm(dim=-1, keepdim=True),
            scales=torch.exp(scales_crop),
            opacities=torch.sigmoid(opacities_crop).squeeze(-1),
            colors=colors_crop,
            viewmats=viewmat,  # [1, 4, 4]
            Ks=K,  # [1, 3, 3]
            width=W,
            height=H,
            tile_size=BLOCK_WIDTH,
            packed=False,
            near_plane=0.01,
            far_plane=1e10,
            render_mode=render_mode,
            sh_degree=sh_degree_to_use,
            sparse_grad=False,
            absgrad=True,
            rasterize_mode=self.config.rasterize_mode,
            # set some threshold to disregrad small gaussians for faster rendering.
            # radius_clip=3.0,
        )
```

So as you see, the radius clip parameter is commented, so it should take the default value which is 0.

I can provide also visual results: the use case is that I want to crop the plate in a car, and replace it with a wall of white gaussians. This is the code that produces the new gaussians

```
# u and v are vectors, it will produce a square of points
# the mask contains which gaussians to keep and which to crop

    new_means = torch.tensor([p1 + t*u + s*v for t in np.linspace(0., 1., 30)  
                                               for s in np.linspace(0.,1.,12)]).to('cuda') 
    means = torch.cat((new_means, params.means[mask]))                 
    params.means = torch.nn.Parameter(means)

    def new_params(shape, value, param, mask= None):
        if mask is None:
            mask = torch.ones(shape, type = bool)
        new_feat = torch.full(shape, value).to('cuda')
        new_feat = torch.cat((new_feat, param[mask]))
        return torch.nn.Parameter(new_feat)

    NEW_COLOR = RGB2SH(1) # white gaussians
    NEW_SCALE = -8
    NEW_OPACITY = 15.1
    
    params.features_dc = new_params(shape= new_means.shape, value= NEW_COLOR,   param= params.features_dc, mask= mask)
    params.features_rest = new_params(shape= (new_means.shape[0], 15, 3),   value= 0., param= params.features_rest, mask= mask)
    params.opacities = new_params(shape= (new_means.shape[0],1),   value= NEW_OPACITY, param= params.opacities, mask= mask)
    params.quats = new_params(shape= (new_means.shape[0], 4), value= 0.,           param= params.quats, mask= mask)
    params.scales = new_params(shape= new_means.shape, value= NEW_SCALE,   param= params.scales, mask= mask)
```
   
Visual results in Supersplat:
I get closer:
![Screenshot from 2024-07-01 10-07-51](https://github.com/nerfstudio-project/gsplat/assets/66904987/3c89af5c-721e-455c-bef7-f83955be5bf2)
I get further:
![Screenshot from 2024-07-01 10-07-42](https://github.com/nerfstudio-project/gsplat/assets/66904987/84e58247-8f51-47ff-8d00-40f7a586a5c4)

Rendering with ns-render:
![00000](https://github.com/nerfstudio-project/gsplat/assets/66904987/95c85c04-4c16-4ac5-9aa5-edcc0f1c3fa1)

**Comment by Atticuszz** [comment]
![image](https://github.com/user-attachments/assets/8c9425f2-ce1c-4f21-9f4b-3a8f440ffa81)
![image](https://github.com/user-attachments/assets/f894453f-d913-4751-b48e-041f22e0c30c)
so wired! as i called `rasterization`,it ocurred a large black fog suddenly! 

```python
    def forward(
        self,
        camtoworlds: Tensor,
        Ks: Tensor,
        width: int,
        height: int,
        render_mode: str = "RGB+ED",
    ):
        assert self.means3d.shape[0] == self.opacities.shape[0]
        opacities = torch.sigmoid(self.opacities)
        colors = torch.cat([self.sh0, self.shN], 1)
        # colors = torch.sigmoid(self.colors)
        scales = torch.exp(self.scales)

        render_colors, render_alphas, info = rasterization(
            means=self.means3d,
            quats=self.quats,
            scales=scales,
            opacities=opacities,
            colors=colors,
            sh_degree=self.config.sh_degree,
            viewmats=torch.linalg.inv(camtoworlds),
            Ks=Ks,
            width=width,
            height=height,
            packed=self.config.packed,
            absgrad=self.config.absgrad,
            sparse_grad=self.config.sparse_grad,
            far_plane=self.config.far_plane,
            near_plane=self.config.near_plane,
            render_mode=render_mode,
            rasterize_mode="classic",
        )

        return render_colors, render_alphas, info
``` we do not change any params but camtoworlds,the image took by `viewer` ,in current pose `rasterization` got almost all black color and almost zero depth

**Comment by Atticuszz** [comment]
> ![image](https://github.com/user-attachments/assets/8c9425f2-ce1c-4f21-9f4b-3a8f440ffa81)
> ![image](https://github.com/user-attachments/assets/f894453f-d913-4751-b48e-041f22e0c30c)
> so wired! as i called `rasterization`,it ocurred a large black fog suddenly! 
> 
> ```python
>     def forward(
>         self,
>         camtoworlds: Tensor,
>         Ks: Tensor,
>         width: int,
>         height: int,
>         render_mode: str = "RGB+ED",
>     ):
>         assert self.means3d.shape[0] == self.opacities.shape[0]
>         opacities = torch.sigmoid(self.opacities)
>         colors = torch.cat([self.sh0, self.shN], 1)
>         # colors = torch.sigmoid(self.colors)
>         scales = torch.exp(self.scales)
> 
>         render_colors, render_alphas, info = rasterization(
>             means=self.means3d,
>             quats=self.quats,
>             scales=scales,
>             opacities=opacities,
>             colors=colors,
>             sh_degree=self.config.sh_degree,
>             viewmats=torch.linalg.inv(camtoworlds),
>             Ks=Ks,
>             width=width,
>             height=height,
>             packed=self.config.packed,
>             absgrad=self.config.absgrad,
>             sparse_grad=self.config.sparse_grad,
>             far_plane=self.config.far_plane,
>             near_plane=self.config.near_plane,
>             render_mode=render_mode,
>             rasterize_mode="classic",
>         )
> 
>         return render_colors, render_alphas, info
> ``` 
> we do not change any params but camtoworlds,the image took by `viewer` ,in current pose `rasterization` got almost all black color and almost zero depth
> ![image](https://github.com/user-attachments/assets/d6037353-6217-4adc-ae65-00e5453c0981)
> 

there is very strange dark point just before camera.I guess Project depth to 3d may have issues
